File: jar/src/main/java/jline/api/qsys/Qsys_mg1_lrpt.java
/**
 * @file M/G/1 queueing system analysis with LRPT scheduling
 *
 * Implements analysis of M/G/1 queues with Longest Remaining Processing Time (LRPT)
 * scheduling. Under LRPT, jobs with the longest remaining time share the processor.
 *
 * @since LINE 3.0
 */
package jline.api.qsys;

import jline.io.Ret;
import jline.util.matrix.Matrix;

public final class Qsys_mg1_lrpt {
    private Qsys_mg1_lrpt() {}

    /**
     * Analyzes an M/G/1 queueing system with LRPT (Longest Remaining Processing Time) scheduling.
     */
    public static Ret.qsys_prio qsys_mg1_lrpt(Matrix lambda, Matrix mu, Matrix cs) {
        double[] lambdaArr = lambda.toArray1D();
        double[] muArr = mu.toArray1D();
        double[] csArr = cs.toArray1D();

        if (lambdaArr.length != muArr.length || lambdaArr.length != csArr.length) {
            throw new IllegalArgumentException("lambda, mu, and cs must have the same length");
        }

        int K = lambdaArr.length;

        for (int i = 0; i < K; i++) {
            if (!(lambdaArr[i] > 0.0)) throw new IllegalArgumentException("lambda[" + i + "] must be positive");
            if (!(muArr[i] > 0.0)) throw new IllegalArgumentException("mu[" + i + "] must be positive");
            if (!(csArr[i] >= 0.0)) throw new IllegalArgumentException("cs[" + i + "] must be non-negative");
        }

        // Compute utilizations
        double[] rhoI = new double[K];
        double rhoTotal = 0.0;
        for (int i = 0; i < K; i++) {
            rhoI[i] = lambdaArr[i] / muArr[i];
            rhoTotal += rhoI[i];
        }

        if (!(rhoTotal < 1.0)) {
            throw new IllegalStateException("System is unstable: utilization rho = " + rhoTotal + " >= 1");
        }

        // Compute overall second moment of service time
        double lambdaTotal = 0.0;
        for (double l : lambdaArr) lambdaTotal += l;
        double[] p = new double[K];
        for (int i = 0; i < K; i++) p[i] = lambdaArr[i] / lambdaTotal;

        double EX2 = 0.0;
        for (int i = 0; i < K; i++) {
            double ES2i = (1.0 + csArr[i] * csArr[i]) / (muArr[i] * muArr[i]);
            EX2 += p[i] * ES2i;
        }

        // The reference has TWO branches and only the exponential one was ported
        // here, so a class with cs != 1 was answered with the exponential
        // formula: on lambda=[0.3,0.2], mu=[1,0.5], cs=[1,sqrt(0.5)] that read
        // W = [13.33, 5.33] against MATLAB's [7.111, 3.333].
        double[] W;
        boolean allExp = true;
        for (int i = 0; i < K; i++) {
            if (Math.abs(csArr[i] - 1.0) >= 1e-6) {
                allExp = false;
                break;
            }
        }
        if (allExp) {
            W = new double[K];
            double term2 = lambdaTotal * EX2 / (2.0 * (1.0 - rhoTotal) * (1.0 - rhoTotal));
            for (int k = 0; k < K; k++) {
                W[k] = (1.0 / muArr[k]) / (1.0 - rhoTotal) + term2;
            }
        } else {
            W = lrptGeneral(lambdaArr, muArr);
        }

        // Compute rhohat
        double Q = 0.0;
        for (int i = 0; i < K; i++) {
            Q += lambdaArr[i] * W[i];
        }
        double rhohat = Q / (1.0 + Q);

        return new Ret.qsys_prio(new Matrix(W), rhohat);
    }

    /**
     * The general (non-exponential) branch: MATLAB's class-based preemptive
     * priority surrogate, with the classes ordered by DECREASING mean service
     * time, which is the order LRPT serves them in.
     *
     * <pre>
     *   W_q(k) = (sum_{j&lt;=k} lambda_j/mu_j^2) / ((1 - rho_{&lt;k})(1 - rho_{&lt;=k}))
     *   W(k)   = W_q(k) + 1/mu_k
     * </pre>
     *
     * <p>The sort must be STABLE, as MATLAB's descending sort is, or two classes
     * of equal mean size swap places and the cumulative sums change.
     */
    private static double[] lrptGeneral(double[] lambdaArr, double[] muArr) {
        int K = lambdaArr.length;
        Integer[] idx = new Integer[K];
        for (int i = 0; i < K; i++) {
            idx[i] = Integer.valueOf(i);
        }
        final double[] meanService = new double[K];
        for (int i = 0; i < K; i++) {
            meanService[i] = 1.0 / muArr[i];
        }
        java.util.Arrays.sort(idx, new java.util.Comparator<Integer>() {
            @Override
            public int compare(Integer a, Integer b) {
                return Double.compare(meanService[b.intValue()], meanService[a.intValue()]);
            }
        });

        double[] W = new double[K];
        double rhoPrev = 0.0;
        double ERk = 0.0;
        for (int k = 0; k < K; k++) {
            int r = idx[k].intValue();
            double rhoCurr = rhoPrev + lambdaArr[r] / muArr[r];
            ERk += lambdaArr[r] / (muArr[r] * muArr[r]);
            W[r] = ERk / ((1.0 - rhoPrev) * (1.0 - rhoCurr)) + 1.0 / muArr[r];
            rhoPrev = rhoCurr;
        }
        return W;
    }
}

File: jar/src/main/kotlin/jline/api/qsys/Qsys_mg1_lrpt.kt
/**
 * @file M/G/1 queueing system analysis with LRPT scheduling
 *
 * Implements analysis of M/G/1 queues with Longest Remaining Processing Time (LRPT)
 * scheduling. Under LRPT, jobs with the longest remaining time share the processor.
 *
 * For LRPT, the expected slowdown for a job of size x is:
 *   E[S(x)]^LRPT = 1/(1-rho) + lambda*E[X^2] / (2*x*(1-rho)^2)
 *
 * Therefore the expected response time is:
 *   E[T(x)]^LRPT = x/(1-rho) + lambda*E[X^2] / (2*(1-rho)^2)
 *
 * LRPT is "Always Unfair" - E[S(y)]^LRPT > 1/(1-rho) for all finite job sizes y.
 * The slowdown is monotonically decreasing in x, converging to 1/(1-rho).
 *
 * References:
 *   - A. Wierman and M. Harchol-Balter, "Classifying scheduling policies with
 *     respect to unfairness in an M/GI/1", SIGMETRICS 2003, Section 3.2.
 *   - M. Harchol-Balter, K. Sigman, and A. Wierman, "Asymptotic convergence
 *     of scheduling policies with respect to slowdown", Performance Evaluation, 2002.
 *
 * @since LINE 3.0
 */
package jline.api.qsys

import jline.io.Ret.qsys_prio
import jline.util.matrix.Matrix

/**
 * Analyzes an M/G/1 queueing system with LRPT (Longest Remaining Processing Time) scheduling.
 *
 * Under LRPT, jobs with the longest remaining processing time share the processor
 * evenly. This prioritizes large jobs over small jobs.
 *
 * @param lambda Matrix (column vector) of arrival rates per class.
 * @param mu     Matrix (column vector) of service rates per class.
 * @param cs     Matrix (column vector) of coefficients of variation per class.
 * @return qsys_prio containing:
 *   - W (Matrix): Vector of mean response times per class
 *   - rho (Double): Overall system utilization
 * @throws IllegalArgumentException if inputs have different lengths or non-positive values
 * @throws IllegalStateException if system is unstable (rho >= 1)
 */
fun qsys_mg1_lrpt(lambda: Matrix, mu: Matrix, cs: Matrix): qsys_prio {
    val lambdaArr = lambda.toArray1D()
    val muArr = mu.toArray1D()
    val csArr = cs.toArray1D()

    require(lambdaArr.size == muArr.size && lambdaArr.size == csArr.size) {
        "lambda, mu, and cs must have the same length"
    }

    val K = lambdaArr.size

    for (i in 0 until K) {
        require(lambdaArr[i] > 0.0) { "lambda[$i] must be positive" }
        require(muArr[i] > 0.0) { "mu[$i] must be positive" }
        require(csArr[i] >= 0.0) { "cs[$i] must be non-negative" }
    }

    // Compute utilizations
    val rhoI = DoubleArray(K) { i -> lambdaArr[i] / muArr[i] }
    val rhoTotal = rhoI.sum()

    check(rhoTotal < 1.0) { "System is unstable: utilization rho = $rhoTotal >= 1" }

    // Compute overall second moment of service time
    // E[X^2] = sum_i (lambda_i / lambda_total) * E[S_i^2]
    val lambdaTotal = lambdaArr.sum()
    val p = DoubleArray(K) { i -> lambdaArr[i] / lambdaTotal }

    var EX2 = 0.0
    for (i in 0 until K) {
        // E[S_i^2] = (1 + cs_i^2) / mu_i^2
        val ES2i = (1.0 + csArr[i] * csArr[i]) / (muArr[i] * muArr[i])
        EX2 += p[i] * ES2i
    }

    // Compute response times using LRPT formula
    // E[T(x)] = x/(1-rho) + lambda*E[X^2] / (2*(1-rho)^2)
    val W = DoubleArray(K)

    val term2 = lambdaTotal * EX2 / (2.0 * (1.0 - rhoTotal) * (1.0 - rhoTotal))

    for (k in 0 until K) {
        val x = 1.0 / muArr[k]  // Mean service time
        val term1 = x / (1.0 - rhoTotal)
        W[k] = term1 + term2
    }

    // Compute rhohat
    var Q = 0.0
    for (i in 0 until K) {
        Q += lambdaArr[i] * W[i]
    }
    val rhohat = Q / (1.0 + Q)

    return qsys_prio(Matrix(W), rhohat)
}

/**
 * Queueing system M/G/1 with LRPT scheduling algorithms.
 */
@Suppress("unused")
class QsysMg1LrptAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}
